src/osd/OSD.cc

```diff
@@ -6418,6 +6425,37 @@ void OSD::do_command(Connection *con, ceph_tid_t tid, vector<string>& cmd, buffe
     ss << "compacted omap in " << duration << " seconds";
   }
 
+  else if (prefix == "smart") {
+    dout(1) << "smartctl is called" << dendl;
+    set<string> devnames;
+    store->get_devices(&devnames);
+    uint64_t smart_timeout = cct->_conf->get_val<uint64_t>("osd_smart_report_timeout");
+    std::string result;
+    
+    json_spirit::mObject json_map; // == typedef std::map<std::string, mValue> mObject;
+    json_spirit::mValue smart_json;
+
+    for (auto it = devnames.begin(); it != devnames.end(); ++it) { 
+      // smartctl works only on physical devices; filter out any non-physical device
+      if (it->find("dm-") != std::string::npos)
```

Comment:
if (dev.find("dm-") == 0) {
  continue;
}

Reply:
and you can replace *it with dev through the rest of this too.  the c++11 for loop style is much more concise :)

Reply:
sleek, thanks!

just double checking that it should be '!=': 
if (dev.find("dm-") != 0) {
continue;
}

Reply:
in this case, ==.. find() returns the position, and we're checking of "dm-" is the prefix of the string starting at offset 0

Reply:
oops, right, sorry!
is there a possibility that "dm-" itself will have another prefix in the device list?

Reply:
i don't think so!

Reply:
great, so we got that covered.
still need to take care of RAID devices which may not return SMART data.